### PR TITLE
internal/dag: Add namespace selector check for Gateways

### LINKED ISSUE #3401: Implement the RouteBindingSelector
A RouteBindingSelector defines a schema for associating routes with the Gateway. If Namespaces and Selector are defined, only routes matching both selectors are associated with the Gateway.

### PR DESCRIPTION
**Pull request by stevesloka**
Filter the HTTPRoutes that match against the namespace field on any struct
looking for if they match the "Same", "All", or "Selector" types.

Fixes #3401

Signed-off-by: Steve Sloka <slokas@vmware.com>

### CODE REVIEW

File: internal/dag/builder_test.go
Comment (skriss) [inline internal/dag/builder_test.go:98]: I'd prefer to use the constants here and in the other gateway defs: https://github.com/kubernetes-sigs/gateway-api/blob/master/apis/v1alpha1/gateway_types.go#L448-L455

File: internal/dag/gatewayapi_processor.go
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:115]: The fact that the key here has a `Namespace: "default"` is confusing because `Namespaces` are cluster-scoped. I'm seeing that it's due to `k8s.NamespacedNameOf(...)` automatically setting a namespace of "default" if it's not provided. I'm thinking we should change the `namespaces` map in the DAG cache to be keyed just on a string (the name), rather than a `NamespacedName`, and not use `NamespacedNameOf` when getting/setting. This looks like the pattern that `IngressClass` uses, which is also cluster-scoped.

File: internal/dag/builder_test.go
Comment (stevesloka) [reply internal/dag/builder_test.go]: 👍

File: internal/dag/gatewayapi_processor.go
Comment (stevesloka) [reply internal/dag/gatewayapi_processor.go:115]: Yup good call! It didn't feel right when I did that, but this makes more sense. I opened https://github.com/projectcontour/contour/pull/3473 to cover.

Comment (youngnick) [reply internal/dag/gatewayapi_processor.go:115]: There is some discussion upstream about making both GatewayClass and IngressClass either namespace-scoped, or have namespace-scoped variants.

**Comment by skriss** [comment]
ready for a rebase since #3473 has merged

File: internal/dag/gatewayapi_processor.go
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:81]: nit: `matchingRoutes` would be a more accurate name for this slice than `validRoutes`, I think.

**Review by skriss — CHANGES_REQUESTED** [review]
one more nit, I think LGTM otherwise (pending rebase) - will check once more when it's updated.

File: internal/dag/gatewayapi_processor.go
Comment (stevesloka) [reply internal/dag/gatewayapi_processor.go]: 👍

**Comment by stevesloka** [comment]
@skriss all rebased and fixed up, please take a look when you get a second.

**Review by skriss — APPROVED** [review]
LGTM

**Comment by skriss** [comment]
now has a conflict to resolve

File: internal/dag/gatewayapi_processor.go
Comment (youngnick) [inline internal/dag/gatewayapi_processor.go:91]: nit: maybe `matchedRoute` instead of `validRoute`?

**Review by youngnick — APPROVED** [review]
LGTM with one tiny nit. Don't block this on that though.

**Pull request merged**
